### PR TITLE
checkTypeEquivalence now considers ProbeInfo

### PR DESCRIPTION
**Pull request by mwachs5**
### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- API modification

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference).

#### Release Notes

Now checkTypeEquivalence will check whether data have the same probe type modifier including writeability and color (layer).

### Reviewer Checklist (only modified by reviewer)
- [x] Did you add the appropriate labels? (Select the most appropriate one based on the "Type of Improvement")
- [x] Did you mark the proper milestone (Bug fix: `3.6.x`, `5.x`, or `6.x` depending on impact, API modification or big change: `7.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [x] Did you do one of the following when ready to merge:
  - [x] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by mwachs5** [comment]
not sure what's going on with the docs change, will revert that.

File: src/test/scala/chisel3/TypeEquivalenceSpec.scala [hunk 1]
```diff
@@ -224,45 +241,94 @@ class TypeEquivalenceSpec extends AnyFlatSpec {
...
+  }
+
+  it should "support RWProbe" in {
+    RWProbe(Bool()).findFirstTypeMismatch(RWProbe(Bool()), true, true, true) should be(None)
+  }
+
+  it should "detect differences between Probe and Not-Probe" in {
+    Probe(Bool()).findFirstTypeMismatch(Bool(), true, true, true) should be(
+      Some(
+        ": Left (Bool with probeInfo: Some(writeable=false, color=None)) and Right (Bool with probeInfo: (None) have different probeInfo."
```
Comment (jackkoenig) [inline src/test/scala/chisel3/TypeEquivalenceSpec.scala:299]:
```suggestion
        ": Left (Bool with probeInfo: Some(writeable=false, color=None)) and Right (Bool with probeInfo: (None)) have different probeInfo."
```
This seems to be missing a parentheses.

File: core/src/main/scala/chisel3/Data.scala [hunk 2]
```diff
@@ -520,63 +520,81 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
...
+    strictWidths:    Boolean,
+    strictProbeInfo: Boolean
   ): Option[String] = {
+
+    def checkProbeInfo(left: Data, right: Data): Option[String] =
+      Option.when(strictProbeInfo && (left.probeInfo != right.probeInfo)) {
+        def probeInfoStr(info: Option[ProbeInfo]) = info.map { info =>
+          s"Some(writeable=${info.writable}, color=${info.color})"
+        }.getOrElse("None")
+        s": Left (${left} with probeInfo: ${probeInfoStr(left.probeInfo)}) and Right ($right with probeInfo: (${probeInfoStr(right.probeInfo)}) have different probeInfo."
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/Data.scala:552]:
```suggestion
        s": Left ($left with probeInfo: ${probeInfoStr(left.probeInfo)}) and Right ($right with probeInfo: (${probeInfoStr(right.probeInfo)})) have different probeInfo."
```
Add missing parentheses and remove unneeded curly braces.

**Review by jackkoenig — APPROVED** [review]
LGTM but please fix parentheses

File: src/test/scala/chisel3/TypeEquivalenceSpec.scala [hunk 3]
```diff
@@ -224,45 +241,94 @@ class TypeEquivalenceSpec extends AnyFlatSpec {
...
+
+  it should "detect differences in presence of probe colors" in {
+    Probe(Bool()).findFirstTypeMismatch(Probe(Bool(), Green), true, true, true) should be(
+      Some(
+        ": Left (Bool with probeInfo: Some(writeable=false, color=None)) and Right (Bool with probeInfo: (Some(writeable=false, color=Some(Green))) have different probeInfo."
+      )
+    )
+  }
+
+  it should "detect differences in probe colors" in {
```
Comment (dtzSiFive) [inline src/test/scala/chisel3/TypeEquivalenceSpec.scala:326]: Can this be checked with bundles or some aggregate? Implementation looks like it should work but good to put that under test.

Same bundle/vector, different modifier-s set on an element or similar?

**Review by dtzSiFive — APPROVED** [review]
Thanks! LGTM!

File: src/test/scala/chisel3/TypeEquivalenceSpec.scala (hunk 3 above)
Comment (mwachs5) [reply src/test/scala/chisel3/TypeEquivalenceSpec.scala:326]: ah gotcha. Yeah will add that!

Comment (mwachs5) [reply src/test/scala/chisel3/TypeEquivalenceSpec.scala:348]: added more bundle-y tests in https://github.com/chipsalliance/chisel/pull/4064/commits/373a61dd9eec1a5f7eb583df4e90be8115dd9a04

Comment (mwachs5) [reply src/test/scala/chisel3/TypeEquivalenceSpec.scala:348]: and vector tests too in https://github.com/chipsalliance/chisel/pull/4064/commits/6fd89a362ff3cf78085e50f96ab847e5c219f604

**Pull request merged**